### PR TITLE
Lead Conversion Error When User is Not Lead Owner In Private Account Sharing Model

### LINKED ISSUE #5810: Lead Conversion Error When User is Not Lead Owner In Private Account Sharing Model
When a customer has their Account Sharing set to “Private” and a User attempts to use the NPSP Lead Conversion tool on a Lead they don’t own, they receive the following error:

_List has no rows for assignment to SObject_

The error occurs in [LD_LeadConvertOverride_CTRL](https://github.com/SalesforceFoundation/NPSP/blob/a91e66e6bfdf55ce0e3d3dedbe770ec3483dee3f/src/classes/LD_LeadConvertOverride_CTRL.cls) when we run the following query on line 436

```
12:24:09.400 (1404300621)|SOQL_EXECUTE_BEGIN|[436]|Aggregations:0|select id, firstname, lastname, npo02__Household__c, MailingStreet, MailingCity, MailingState, MailingPostalCode, MailingCountry, MailingLatitude, MailingLongitude, HomePhone, Email from contact where id= '003xxxxxxxxxxxxx
12:24:09.400 (1416680139)|SOQL_EXECUTE_END|[436]|Rows:0
```

This occurs after we convert the Lead and then query for the Contact. They don’t have access to the Contact since we create the Contact and Account with the Owner Id set to the Lead’s Owner Id. 

The core Lead Conversion tool does not run into errors when converting Leads for different Owners despite setting the Contact/Account Owner to that of the Lead’s Owner. 

### Temporary Workarounds:
1. On the Lead Conversion screen, switch the Owner to the User converting the Lead.
2. Determine if there are any modifications that can be made to Sharing Rules. 

### Steps to Repeat:
In scratch org:

* Go to Setup → Object Manager → Lead → Buttons, Links, and Actions. Confirm there is an override on the Convert button to: LD_LeadConvertOverride.
* Clone Standard User Profile
    * add Create/Read/Update/Delete for Lead, Contact, and Account (All Fields: Read/Write)
    * Make the Organization Record Type the default Account Record Type for that Profile.
    * Add LD_LeadConvertOverride to Visualforce Page Access
* Create new User and assign to new Profile.
* Go to Setup → Sharing Settings and set Account access to Private.

**Tests:**

_**Test 1:**_
Create Lead as System Admin. Use Lead Conversion tool as System Admin.
Confirm Contact/Account are created successfully.

**_Test 2:_**
Create Lead as System Admin. Use Lead Conversion tool as non-System Admin.
Confirm Error at top of conversion page:
“Error: List has no rows for assignment to SObject.”
Screen Shot 2020-09-03 at 3 24 47 PM

**_Test 3:_**
Create Lead As non-System Admin. Use Lead Conversion tool as non-System Admin.
Confirm no error.

**_Test 4:_**
Switch lead conversion to core lead conversion (i.e. remove the override on the Lead Convert button). Create Lead as System Admin. Convert Lead as non-System Admin.

### PR DESCRIPTION
**Pull request by dfuller114**
[W-8049112](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07B0000008cmv7IAA/view)
# Critical Changes

# Changes

# Issues Closed

Fixes #5810 

# Community Ideas Delivered

# Features Intended for Future Release

# Features for Elevate Customers

# New Metadata

# Deleted Metadata

### CODE REVIEW

File: force-app/main/default/classes/LD_LeadConvertOverride_CTRL.cls [hunk 1]
```diff
@@ -141,17 +141,17 @@ public with sharing class LD_LeadConvertOverride_CTRL {
     /*******************************************************************************************************
     * @description Checks the value of the 'Company' field for values associated with 1x1 or Individual Accounts
     */
-    public Boolean isSpecialAccount {
+    public Boolean isLegacyAccount {
```
[the remark this answers is not in the document]
Comment (dfuller114) [reply force-app/main/default/classes/LD_LeadConvertOverride_CTRL.cls]: good callout. We actually use `LegacyHousehold` in our domain so I pushed a change to update this getter.

File: force-app/main/default/classes/LD_LeadConvertOverride_TEST.cls [hunk 2]
```diff
@@ -38,6 +38,42 @@ private class LD_LeadConvertOverride_TEST {
...
+                    'user does not have permission to view the contact record.';
+            if (queryException != null) {
+                queryExceptionMessage += ' Query Exception thrown: ' + queryException.getMessage();
+            }
+            System.assert(queryException == null, queryExceptionMessage);
+        }
+    }
+
+    @IsTest
+    static void newLeadConversion_UserDoesNotOwnLeadRecord() {
```
Comment (lparrott) [inline force-app/main/default/classes/LD_LeadConvertOverride_TEST.cls:72]: Still a TODO?

**Review by andrewyu-salesforce — CHANGES_REQUESTED** [review]

File: force-app/main/default/classes/LD_LeadConvertOverride_CTRL.cls [hunk 3]
```diff
@@ -686,4 +500,314 @@ public with sharing class LD_LeadConvertOverride_CTRL {
...
+            UTIL_DMLService.insertRecord(npspAddress.getRecord());
+        }
+    }
+
+    private void enabledAddressManagement() {
+        UTIL_CustomSettingsFacade.getContactsSettings()
+                .Household_Account_Addresses_Disabled__c = isHouseholdAddressesDisabled();
+        UTIL_CustomSettingsFacade.getContactsSettings()
+                .Organizational_Account_Addresses_Enabled__c = isOrganizationAddressesDisabled();
+    }
```
[the remark this answers is not in the document]
Comment (dfuller114) [reply force-app/main/default/classes/LD_LeadConvertOverride_CTRL.cls]: @andrewyu-salesforce good callout. just pushed an update. let me know what you think

**Review by andrewyu-salesforce — APPROVED** [review]
@dfuller114 everything looks great. Nice refactor

**Review by lparrott — APPROVED** [review]

File: force-app/main/default/classes/LD_LeadConvertOverride_CTRL.cls [hunk 4]
```diff
@@ -686,4 +504,363 @@ public with sharing class LD_LeadConvertOverride_CTRL {
...
+            this.isPrimary = isPrimary;
+        }
+
+        private void createNonDupeAffiliation() {
+            AFFL_Affiliations_TDTM.createNonDupeAffiliation(leadToConvert, contactId, affiliationId, isPrimary);
+        }
+    }
+
+    @TestVisible
+    private with sharing class DMLWithoutSharing {
```
Comment (lparrott) [inline force-app/main/default/classes/LD_LeadConvertOverride_CTRL.cls:784]: Should this be `without sharing`?

File: force-app/main/default/classes/LD_LeadConvertOverride_CTRL.cls [hunk 5]
```diff
@@ -686,4 +504,363 @@ public with sharing class LD_LeadConvertOverride_CTRL {
...
+                System.Label.leadConvertContactSelection));
+        return null;
+    }
+
+    private Boolean isContactSelected() {
+        return convertedContactId != null;
+    }
+
+    @TestVisible
+    private with sharing class AffiliationNonDupeWithoutSharing {
```
Comment (lparrott) [inline force-app/main/default/classes/LD_LeadConvertOverride_CTRL.cls:764]: Should this be `without sharing`?

**Review by andrewyu-salesforce — APPROVED** [review]
great unit test on the user permission

**Pull request merged**